packages/vscode-extension/src/builders/BuildManager.ts

```diff
@@ -104,148 +37,130 @@ export class BuildManager {
-          const hash = await getAppHash(build.appPath);
-
-          if (hash === cacheInfo.buildHash) {
-            Logger.log("Cache hit on iOS build. Using existing build.");
-            return build;
-          }
-        }
-      }
-    } catch (e) {
-      // we only log the error and ignore it to allow new build to start
-      Logger.error("Error while attempting to load cached build", e);
+    const hash = await getAppHash(appPath);
+    const hashesMatch = hash === cacheInfo.buildHash;
+    if (hashesMatch) {
+      Logger.log("Using existing build.");
+      return build;
     }
+  } catch (e) {
+    // we only log the error and ignore it to allow new build to start
+    Logger.error("Error while attempting to load cached build", e);
     return undefined;
   }
+}
 
-  private async startIOSBuild(
-    deviceInfo: IOSDeviceInfo,
-    forceCleanBuild: boolean,
-    cancelToken: CancelToken,
-    progressListener: (newProgress: number) => void
-  ): Promise<IOSBuildResult> {
-    const newFingerprint = await generateWorkspaceFingerprint();
-    if (!forceCleanBuild) {
-      const buildResult = await this.loadIOSCachedBuild(newFingerprint);
-      if (buildResult) {
-        return buildResult;
-      }
-    } else {
-      // we reset the cache when force clean build is requested as the newly started build may end up being cancelled
-      extensionContext.workspaceState.update(IOS_BUILD_CACHE_KEY, undefined);
-    }
-    this.buildOutputChannel = window.createOutputChannel("React Native IDE (iOS build)", {
-      log: true,
-    });
-    const build = await buildIos(
-      deviceInfo,
-      getAppRootFolder(),
-      forceCleanBuild,
-      cancelToken,
-      this.buildOutputChannel,
-      progressListener,
-      () => {
-        return this.dependencyManager.checkPodsInstalled();
-      },
-      (appRootFolder: string, cleanBuild: boolean, token: CancelToken) => {
-        return this.dependencyManager.installPods(appRootFolder, cleanBuild, token);
-      }
-    );
-    const buildResult = { ...build, platform: DevicePlatform.IOS };
-
-    // store build info in the cache
-    const newBuildHash = await getAppHash(build.appPath);
-    const buildInfo = { fingerprint: newFingerprint, buildHash: newBuildHash, buildResult };
-    extensionContext.workspaceState.update(IOS_BUILD_CACHE_KEY, buildInfo);
+export async function didFingerprintChange(platform: DevicePlatform) {
+  const newFingerprint = await generateWorkspaceFingerprint();
+  const { fingerprint } = getCachedBuild(platform) ?? {};
 
-    return { ...build, platform: DevicePlatform.IOS };
-  }
+  return newFingerprint !== fingerprint;
 }
 
 async function getAppHash(appPath: string) {
   return (await calculateMD5(appPath)).digest("hex");
 }
+
+function getAppPath(build: BuildResult) {
+  return build.platform === DevicePlatform.Android ? build.apkPath : build.appPath;
+}
+
+const ANDROID_BUILD_CACHE_KEY = "android_build_cache";
```

Comment:
> as a rule to avoid duplication
> I would treat it as "the gun is always loaded" type of thing.

I don't understand this argument at all, especially that second sentence.

These constants are used to fetch things from the cache and the only things that should use it are two functions below it. Reducing scope of the variables actually improves readability because now you are sure that they're not used anywhere above the definition.

Reply:
I don't agree, reading the line: `const IOS_BUILD_CACHE_KEY = "ios_build_cache";`
does not provide any new information to the reader, but if anybody wanted to store some constant and would always do it on top of the file (as you usually do anyway). They could see that the constant is already defined and they should use or change the old one. I see your point that in this case it make little if any difference and I said as much in my original message, I just think in most cases it is beneficial to define magic constants like that on top of the file and I wouldn't like to break that rule here, a specially that I really believe that it does not improve readability.

Reply:
> I don't agree, reading the line: const IOS_BUILD_CACHE_KEY = "ios_build_cache";
does not provide any new information to the reader

That's not my point. Also, if anybody wants to add a new constant, they're free to put it anywhere they see fit. As I said, in this case it prevents from using it directly and forcing to use accessor functions.

In the future, I'd like to keep all workspace store operations in one file since it's effectively global persistent storage in the project.